rasa/engine/validation.py

```diff
@@ -0,0 +1,324 @@
+            f"method '{node.fn}' does not have a type annotation for "
+            f"its return value. Type annotations are required for all graph "
+            f"components to validate the graph's structure."
+        )
+
+    if is_training and not isinstance(return_type, Fingerprintable):
+        raise GraphSchemaValidationException(
+            f"Node '{node_name}' uses a component '{node.uses.__name__}' whose method "
+            f"'{node.fn}' does not return a fingerprintable "
+            f"output. This is required for caching. Please make sure you're "
+            f"using a return type which implements the "
+            f"'{Fingerprintable.__name__}' protocol."
+        )
+
+
+def _validate_types_of_reserved_keywords(
+    params: Dict[Text, ParameterInfo], node_name: Text, node: SchemaNode, fn_name: Text
+) -> None:
+    for param_name, param in params.items():
+        if param_name in _keywords_expected_types:
+            if not typing_utils.issubtype(
+                param.type_annotation, _keywords_expected_types[param_name]
+            ):
+                raise GraphSchemaValidationException(
+                    f"Node '{node_name}' uses a graph component "
+                    f"'{node.uses.__name__}' which has an incompatible type "
+                    f"'{param.type_annotation}' for the '{param_name}' parameter in "
+                    f"its '{fn_name}' method. Expected type "
+                    f"'{ _keywords_expected_types[param_name]}'."
+                )
+
+
+def _validate_constructor(
+    node_name: Text, node: SchemaNode, create_fn_params: Dict[Text, ParameterInfo],
+) -> None:
+    _validate_types_of_reserved_keywords(
+        create_fn_params, node_name, node, node.constructor_name
+    )
+
+    for param_name in _required_args(create_fn_params):
+        if node.eager:
+            raise GraphSchemaValidationException(
+                f"Node '{node_name}' has a constructor which has a "
+                f"required parameter '{param_name}'. Extra parameters can only "
+                f"supplied to be the constructor if the node is being run "
+                f"in lazy mode."
+            )
+        if not node.eager and param_name not in node.needs:
+            raise GraphSchemaValidationException(
+                f"Node '{node_name}' uses a component '{node.uses.__name__}' which "
+                f"needs the param '{param_name}' to be provided to its method "
+                f"'{node.constructor_name}'. Please make sure to specify the "
+                f"parameter in the node's 'needs' section."
+            )
+
+
+def _validate_needs(
+    node_name: Text,
+    node: SchemaNode,
+    graph: GraphSchema,
+    create_fn_params: Dict[Text, ParameterInfo],
+    run_fn_params: Dict[Text, ParameterInfo],
+) -> None:
+    available_args, has_kwargs = _get_available_args(
+        node, create_fn_params, run_fn_params
+    )
+
+    for param_name, parent_name in node.needs.items():
```

Comment:
What about if we have an "available_args" which is not provided by the "needs"? Do we not need to check in that direction too?

Reply:
That's done in `_validate_constructor` and `_validate_run_fn`

Reply:
ah yes - ofc!